Add improvements to working with errors documentation. (#2467)
wso2/docs-is: en/docs/guides/identity-lifecycles/account-locking-by-failed-login-attempts.md

# Account Locking by Failed Login Attempts

WSO2 Identity Server can be configured to lock a user account when a
number of consecutive failed login attempts are exceeded. First, you
need to configure WSO2 Identity Server for user account locking and
disabling. The following section explain how to configure this.

> ### Configuring WSO2 IS for account locking

Reviewer: ```suggestion
## Configure WSO2 IS for account locking
```

1.  Ensure that the identity listener
    with the `            priority=50           ` is set to **false**
    and the identity listener with the `            priority=95           ` is set to **true**
    in the
    `            <IS_HOME>/repository/conf/deployment.toml           `
    file by adding the following configuration.  

    !!! info 
        This is already configured this way by default. You can skip this
        step if you have not changed this configuration previously.

    ??? note "Click to see the code block"

        ``` xml
        [event.default_listener.identity_mgt]
        priority= "50"
        enable = false
        [event.default_listener.governance_identity_mgt]
        priority= "95"
        enable = true
        ```
    

> 2.  <a name= "failedloginattempts"></a> Start the Identity Server and log into the management console using

Reviewer: ```suggestion
2.  <a name= "failedloginattempts"></a> Start the Identity Server and log in to the management console using
```

    your tenant credentials.

    !!! tip
    
        **Alternatively**, you can also use the
        `                         IdentityGovernanceAdminService                       `
        SOAP service to do this instead of using the management console UI.

>         See [Calling Admin Services](../../develop/calling-admin-services) for more

Reviewer: We need to change these links according to the new structure.

        information on how to invoke this SOAP service. If you are using the
        SOAP service to configure this, you do not need to follow the steps
        given below this note.
    

3.  Click **Resident** under **Identity Providers** found in the
    **Main** tab.
4.  Expand the **Login Attempts Security** tab.
5.  Expand the **Account Lock** tab and select the **Lock user accounts** checkbox. Click **Update** to save changes.  

>     ![account-lock-enabled](../assets/img/using-wso2-identity-server/account-lock-enabled.png) 

Reviewer: Make sure this path is correct

    !!! tip
    
        If a user is assigned the **Internal/system** role, the user can
        bypass account locking even if the user exceeds the specified number
        of **Maximum Failed Login Attempts**.
    
        !!! note
                WSO2 Identity Server has the **Internal/system** role configured by
                default. But generally a new user is not assigned the
                **Internal/system** role by default. Required roles can be assigned
                to a user depending on the set of permission a user needs to have.
                For more information on roles and permission, see [Configuring Roles

>                 and Permissions](../../learn/configuring-roles-and-permissions).

Reviewer: Change link

<p class="admonition-title">Tip</p>
Add the following configuration to the <code> &lt;IS_HOME&gt;/repository/conf/deployment.toml </code> file to 
<strong>Enable</strong> Account lock on failed login by default for all the tenants at tenant creation. 
<div class="code panel pdl" style="border-width: 1px;">
    <div class="codeContent panelContent pdl">
    <pre class="html/xml" data-syntaxhighlighter-params="brush: html/xml; gutter: false; theme: Confluence" 
    data-theme="Confluence" style="brush: html/xml; gutter: false; theme: Confluence"><code>[identity_mgt.account_locking]<br>enable_account_locking=true</code></pre>
    </div>
  </div>
 </div>
</div>
</td>
</tr>
</tbody>
</table>

If you want to configure different settings for another tenant, log out
and follow the same steps to configure these properties for the other
tenants.

!!! note "Configuring WSO2 IS for automatic account unlock"
      
    A user account locked by failed login attempts can be unlocked 
    by setting the <strong>Account Unlock Time</strong> period.
    
    If the lock time is set to 0, the account has to be unlocked by an admin
    user. For more information about this, see [Account locking for a

>     particular user](../../learn/locking-a-specific-user-account).

Reviewer: Change link

    

### Sending email notifications for account locking

Once you have configured WSO2 Identity Server for account locking by
failed login attempts, you can also configure the WSO2 IS to send an
email to the user's email address when the user account is locked due to
failed login attempts. To configure this, follow the steps below.

-  Enable the email sending configurations of the WSO2 Identity Server

>     as explained [here](../../setup/configuring-email-sending).

Reviewer: Change link

    !!! tip
    
        The email template used to send the email notification for
        account locking is the **AccountLock** template and the template
        used for account disabling is the **AccountDisable** template. You
        can edit and customize the email template. For more information on
        how to do this, see [Customizing Automated

>         Emails](../../learn/customizing-automated-emails).

Reviewer: Change link

WSO2 Identity Server uses separate email templates for notifying, 

- Account locking by exceeding `Maximum Failed Login Attempts`
- Account unlocking by exceeding `Account Unlock Time`

Add the following email templates by referring to the instructions in 

> [Customizing AutomatedEmails](../../learn/customizing-automated-emails).

Reviewer: Change link

                   </p>
               </td>
           </tr>
           <tr>
               <td style="text-align: left; padding: 30px 50px 50px 50px" valign="top">
                   <p style="font-size: 18px; margin: 0; line-height: 24px; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #505050; text-align: left;">
                       Thanks,<br/>WSO2 Identity Server Team
                   </p>
               </td>
           </tr>
           <tr>
               <td colspan="2" align="center" style="padding: 20px 40px 40px 40px;" bgcolor="#f0f0f0">
                   <p style="font-size: 12px; margin: 0; line-height: 24px; font-family: 'Nunito Sans', Arial, Verdana, Helvetica, sans-serif; color: #777;">
                       &copy; 2018
                       <a href="http://wso2.com/" target="_blank" style="color: #777; text-decoration: none">WSO2</a>
                       <br>
                       787 Castro Street, Mountain View, CA 94041.
                   </p>
               </td>
           </tr>
           </table>
       </td>
       </tr>
       </table>]]>
    ```
    - Footer : ---
    

> ### Configuring WSO2 IS for Failed OTP attempts based account locking

Reviewer: ```suggestion
## Configure WSO2 IS for Failed OTP attempts based account locking
```

WSO2 Identity Server can be configured to lock a user account when the number of consecutive failed OTP attempts is exceeded. 
First, you need to configure the WSO2 Identity Server for user account locking and disabling according to 

> [Configuring WSO2 IS for account locking](../../learn/account-locking-by-failed-login-attempts/#configuring-wso2-is-for-account-locking).

Reviewer: Change link

1. Add the following configurations into `<IS_HOME>/repository/conf/deployment.toml` file to enable account locking for 
each type of OTP attempts.

    - For Email OTP:
    ```
    [authentication.authenticator.email_otp.parameters]
    EnableAccountLockingForFailedAttempts = true
    ```
    
    - For SMS OTP:
    ```
    [authentication.authenticator.sms_otp.parameters]
    EnableAccountLockingForFailedAttempts = true
    ```
   
        !!! note
            Since `BackupCode = true` in the default configuration, configure the backup code claim according to 

>             [Configuring Backup Codes for SMSOTP](../../learn/configuring-sms-otp/#configuring-backup-codes-for-smsotp). 

Reviewer: Change link
